File: backends/java/zappy-application/src/main/java/com/zappymart/application/promotions/CountPromotionUse.java
package com.zappymart.application.promotions;

import com.zappymart.application.ports.DomainEventHandler;
import com.zappymart.application.ports.PromotionRepository;
import com.zappymart.application.ports.UnitOfWork;
import com.zappymart.domain.ordering.OrderPlaced;
import com.zappymart.domain.promotions.PromotionCode;

public final class CountPromotionUse implements DomainEventHandler<OrderPlaced> {

    private final UnitOfWork unitOfWork;
    private final PromotionRepository promotionRepository;

    public CountPromotionUse(UnitOfWork unitOfWork, PromotionRepository promotionRepository) {
        this.unitOfWork = unitOfWork;
        this.promotionRepository = promotionRepository;
    }

    @Override
    public Class<OrderPlaced> eventType() {
        return OrderPlaced.class;
    }

    @Override
    public void handle(OrderPlaced event) {
        if (event.promotionCode() == null) {
            return;
        }
        unitOfWork.inTransaction(() -> {
            PromotionCode.parse(event.promotionCode())
                    .flatMap(promotionRepository::byCode)
                    .ifPresent(promotion -> promotionRepository.save(promotion.afterOneMoreUse()));
            return true;
        });
    }
}

File: backends/kotlin/zappy-application/src/main/kotlin/nl/zappymart/application/promotions/CountPromotionUse.kt
package nl.zappymart.application.promotions

import nl.zappymart.application.ports.PromotionRepository
import nl.zappymart.domain.ordering.OrderPlaced

class CountPromotionUse(private val promotions: PromotionRepository) {

    fun handle(event: OrderPlaced) {
        val code = event.promotionCode ?: return
        val promotion = promotions.findByCode(code) ?: return
        promotions.save(promotion.usedOnce())
    }
}
